java/mtsj/core/src/main/java/io/oasp/application/mtsj/predictionmanagement/dataaccess/impl/dao/PredictionDayDataDaoImpl.java

```diff
@@ -0,0 +1,199 @@
+    entity.setDish(dish);
+    entity.setKey("_date");
+    entity.setValue(startDate.toString());
+    getEntityManager().persist(entity);
+  }
+
+  public void generatePredictionFor(DishEntity dish) {
+
+    // Prepare model parameters
+
+    getEntityManager().createNativeQuery("DELETE FROM TMP_PREDICTION_MODEL").executeUpdate();
+    getEntityManager().createNativeQuery(
+      "INSERT INTO TMP_PREDICTION_MODEL SELECT "
+      + "key, value FROM PREDICTION_ALL_MODELS "
+      + "WHERE idDish = ?1"
+    )
+    .setParameter(1, dish.getId())
+    .executeUpdate();
+
+    // Do prediction
+
+    getEntityManager().createNativeQuery("DELETE FROM TMP_PREDICTION_FORECAST").executeUpdate();
+    getEntityManager().createNativeQuery(
+      "CALL _SYS_AFL.PAL_ARIMA_FORECAST(PREDICTION_FORECAST_DATA, TMP_PREDICTION_MODEL, "
+      + "PREDICTION_ARIMA_PARAMS, TMP_PREDICTION_FORECAST) WITH OVERVIEW"
+    ).executeUpdate();
+
+    // Save prediction
+
+    CriteriaBuilder cb = getEntityManager().getCriteriaBuilder();
+    CriteriaDelete<PredictionDayDataEntity> criteriaDelete = cb.createCriteriaDelete(PredictionDayDataEntity.class);
+    Root<PredictionDayDataEntity> from = criteriaDelete.from(PredictionDayDataEntity.class);
+    criteriaDelete.where(cb.equal(from.get("dish").get("id"), dish.getId()));
+    getEntityManager().createQuery(criteriaDelete).executeUpdate();
+
+    getEntityManager().createNativeQuery(
+      "INSERT INTO PREDICTION_ALL_FORECAST (idDish, dishName, timestamp, forecast) "
+      + "SELECT ?1 AS idDish, Dish.name AS dishName, timestamp, forecast "
+      + "FROM TMP_PREDICTION_FORECAST LEFT JOIN Dish ON Dish.id = ?1"
+    )
+    .setParameter(1, dish.getId())
+    .executeUpdate();
+  }
+
+  @Override
+  public List<PredictionDayDataEntity> getPrediction(PredictionCriteriaEto criteria) {
+
+    List<DishEntity> dishes = getDishDao().findDishs(new DishSearchCriteriaTo()).getResult();
+
+    // delete old forecast data
+    CriteriaBuilder cb = getEntityManager().getCriteriaBuilder();
+    CriteriaDelete<PredictionForecastDataEntity> criteriaDelete = cb.createCriteriaDelete(PredictionForecastDataEntity.class);
+    criteriaDelete.from(PredictionForecastDataEntity.class);
+    getEntityManager().createQuery(criteriaDelete).executeUpdate();
+
+    // add new forecast data
+    for (int i = 0; i < criteria.getTemperatures().size(); i++) {
+      PredictionForecastDataEntity forecastData = new PredictionForecastDataEntity();
+      forecastData.setTimestamp(i+1);
+      forecastData.setTemperature(criteria.getTemperatures().get(i) == null ? 0 : criteria.getTemperatures().get(i));
+      forecastData.setHoliday(criteria.getHolidays().get(i) == null ? 0 : 1);
+      getEntityManager().persist(forecastData);
+    }
+
+    // Training
+
+    for (DishEntity dish : dishes) {
+      train(dish, criteria.getStartBookingdate());
+    }
+
+    // Prediction
+
+    for (DishEntity dish : dishes) {
+      generatePredictionFor(dish);
+    }
```

Comment:
mhm. So we are doing a training and according prediction for each dish. This is triggered online for each prediction? I am a little scared about performance or maybe I just did not fully understand the power of SAP Hana (hey hurray) but would this code perform with multiple concurrent users? Even if SAP Hana takes zero time still Hibernate has to do something and you can not hide the fact that JDBC is involved what causes a TCP overhead for each dish.
To come to the point: Could we not just do all this in a single statement if we are going to do it for each dish anyhow?
I see lots of potential for optimization. If Hana rocks so heavily that we do not have to care about this, I am even more happy to push this ahead but I am not yet convinced that optimizing this code is just a waste of time :)